lsp-inline-completion.el

```diff
@@ -0,0 +1,487 @@
+    map)
+  "Keymap active when showing inline code suggestions")
+
+;;;###autoload
+(defface lsp-inline-completion-overlay-face
+  '((t :inherit shadow))
+  "Face for the inline code suggestions overlay."
+  :group 'lsp-mode)
+
+;; Local Buffer State
+
+(defvar-local lsp-inline-completion--items nil "The completions provided by the server")
+(defvar-local lsp-inline-completion--current nil "The current suggestion to be displayed")
+(defvar-local lsp-inline-completion--overlay nil "The overlay displaying code suggestions")
+(defvar-local lsp-inline-completion--start-point nil "The point where the completion started")
+
+(defcustom lsp-before-inline-completion-hook nil
+  "Hooks run before starting code suggestions"
+  :type 'hook
+  :group 'lsp-mode)
+
+(defcustom lsp-after-inline-completion-hook nil
+  "Hooks executed after asking for code suggestions."
+  :type 'hook
+  :group 'lsp-mode)
+
+(defcustom lsp-inline-completion-accepted-hook nil
+  "Hooks executed after accepting a code suggestion. The hooks receive the
+text range that was updated by the completion"
+  :type 'hook
+  :group 'lsp-mode)
+
+(defcustom lsp-inline-completion-cancelled-hook nil
+  "Hooks executed after cancelling the completion UI"
+  :type 'hook
+  :group 'lsp-mode)
+
+(defcustom lsp-inline-completion-before-show-hook nil
+  "Hooks executed before showing a suggestion."
+  :type 'hook
+  :group 'lsp-mode)
+
+(defcustom lsp-inline-completion-shown-hook nil
+  "Hooks executed after showing a suggestion."
+  :type 'hook
+  :group 'lsp-mode)
+
+(defcustom lsp-inline-completion-overlay-priority 9000
+  "The priority of the overlay."
+  :type '(choice (const :tag "No Priority" nil)
+                 (integer :tag "Simple, Overriding Priority")
+                 (cons :tag "Composite"
+                       (choice (integer :tag "Primary")
+                               (const :tag "Primary Unset" nil))
+                       (integer :tag "Secondary")))
+  :group 'lsp-mode)
+
+(defsubst lsp-inline-completion--overlay-visible ()
+  "Return whether the `overlay' is avaiable."
+  (and (overlayp lsp-inline-completion--overlay)
+       (overlay-buffer lsp-inline-completion--overlay)))
+
+(defun lsp-inline-completion--clear-overlay ()
+  "Hide the suggestion overlay"
+  (when (lsp-inline-completion--overlay-visible)
+    (delete-overlay lsp-inline-completion--overlay))
+  (setq lsp-inline-completion--overlay nil))
+
+
+(defun lsp-inline-completion--get-overlay (beg end)
+  "Build the suggestions overlay"
+  (when (overlayp lsp-inline-completion--overlay)
+    (lsp-inline-completion--clear-overlay))
+
+  (setq lsp-inline-completion--overlay (make-overlay beg end nil nil t))
+  (overlay-put lsp-inline-completion--overlay 'keymap lsp-inline-completion-active-map)
+  (overlay-put lsp-inline-completion--overlay 'priority lsp-inline-completion-overlay-priority)
+
+  lsp-inline-completion--overlay)
+
+
+(defun lsp-inline-completion-show-keys ()
```

Comment:
nit: is this called from somewhere? Should we make this a command instead of just a function?

Reply:
Oh, I was calling this on the `shown-hook` , leaving up to the user to configure it. IMHO it does not make sense to show the completion overlay and not the `Completion 1/N` message, so I've updated the show-overlay function to call it.

Reply:
That sounds good to me. If this is called already, can we make this private function instead of public one?

Reply:
done